rust/src/client.rs

```diff
@@ -0,0 +1,842 @@
+        };
+
+        let cb = cb_data.cb.expect("callback was None");
+        let cb_context = cb_data.cb_context;
+
+        let mut payload = Vec::with_capacity(256);
+        msg.append_to_sbp_buffer(&mut payload);
+        let payload_ptr = payload.as_mut_ptr();
+
+        unsafe {
+            cb(
+                msg.get_sender_id().unwrap_or(0),
+                msg.sbp_size() as u8,
+                payload_ptr,
+                cb_context,
+            )
+        };
+    }
+}
+
+unsafe impl Send for Context<'_> {}
+unsafe impl Sync for Context<'_> {}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+struct Callback {
+    node: usize,
+    msg_type: u16,
+    cb: sbp_msg_callback_t,
+    cb_context: *mut c_void,
+    key: Key,
+}
+
+#[repr(C)]
+struct Event {
+    condvar: parking_lot::Condvar,
+    lock: parking_lot::Mutex<()>,
+}
+
+impl Event {
+    fn new() -> Self {
+        Self {
+            condvar: parking_lot::Condvar::new(),
+            lock: parking_lot::Mutex::new(()),
+        }
+    }
+
+    fn wait_timeout(&self, ms: i32) -> bool {
+        let mut started = self.lock.lock();
+        let result = self
+            .condvar
+            .wait_for(&mut started, Duration::from_millis(ms as u64));
+        !result.timed_out()
+    }
+
+    fn set(&self) {
+        let _ = self.lock.lock();
+        if !self.condvar.notify_one() {
+            eprintln!("event set did not notify anything");
+        }
+    }
+}
+
+#[repr(C)]
+struct Lock(parking_lot::Mutex<()>);
+
+impl Lock {
+    fn new() -> Self {
+        Self(parking_lot::Mutex::new(()))
+    }
+
+    fn lock(&self) -> parking_lot::MutexGuard<()> {
+        self.0.lock()
+    }
+
+    fn acquire(&self) {
+        std::mem::forget(self.0.lock());
+    }
+
+    fn release(&self) {
+        // Safety: Only called via libsettings_unlock after libsettings_lock was called
+        unsafe { self.0.force_unlock() }
+    }
+}
+
+struct CtxPtr(*mut c_void);
+
+unsafe impl Sync for CtxPtr {}
+unsafe impl Send for CtxPtr {}
+
+#[no_mangle]
+unsafe extern "C" fn register_cb(
+    ctx: *mut c_void,
+    msg_type: u16,
+    cb: sbp_msg_callback_t,
+    cb_context: *mut c_void,
+    node: *mut *mut sbp_msg_callbacks_node_t,
+) -> i32 {
+    let context: &mut Context = &mut *(ctx as *mut _);
+    let _guard = context.lock.lock();
+    let ctx_ptr = CtxPtr(ctx);
+    let key = context.link.register_by_id(&[msg_type], move |msg: SBP| {
+        let context: &mut Context = &mut *(ctx_ptr.0 as *mut _);
+        context.callback_broker(msg)
+    });
+    context.callbacks.push(Callback {
+        node: node as usize,
+        msg_type,
+        cb,
+        cb_context,
+        key,
+    });
+    0
+}
+
+#[no_mangle]
+unsafe extern "C" fn unregister_cb(
+    ctx: *mut c_void,
+    node: *mut *mut sbp_msg_callbacks_node_t,
+) -> i32 {
+    let context: &mut Context = &mut *(ctx as *mut _);
+    if (node as i32) == 0 {
+        return -127;
+    }
+    let idx = match context
```

Comment:
I think this use to have the lock held, no?  We're holding it in `register_cb`, so would expect it to be held here too.

```rust
    let _guard = context.lock.lock();
```

Reply:
Oh right, there was also one in the callback broker too that kept failing to acquire the lock. I tried removing them all but missed that one I guess. Strange that it seems to work like this. I should just figure out why the broker one was having problems

Reply:
> 
> 
> Oh right, there was also one in the callback broker too that kept failing to acquire the lock. I tried removing them all but missed that one I guess. Strange that it seems to work like this. I should just figure out why the broker one was having problems

From what I remember, it was necessary in a prior formulation of this code since I thought the read threads where registering and unregistering themselves for callbacks in parallel, when this wasn't happening under guard it causes settings responses to get lost.  If it isn't required now I'm not sure what would've changed to cause that.

Reply:
> 
> 
> > Oh right, there was also one in the callback broker too that kept failing to acquire the lock. I tried removing them all but missed that one I guess. Strange that it seems to work like this. I should just figure out why the broker one was having problems
> 
> From what I remember, it was necessary in a prior formulation of this code since I thought the read threads where registering and unregistering themselves for callbacks in parallel, when this wasn't happening under guard it causes settings responses to get lost. If it isn't required now I'm not sure what would've changed to cause that.

This is my understanding of what's happening
+ 5 thread get spawned in client::read_all
+ each thread kicks off `settings_read_by_idx` C call for settings read by index response
+ C call calls `setting_sbp_cb_register`
   - This is already looking racy because not we have 2+ threads that aren't under lock looking over data in the context, looking for the "last" cb node: https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/setting_sbp_cb.c#L342
   - `setting_sbp_cb_register` should probably hold the lock while it's modifying data in ctx
+ A callback is built up and sent to Rust land: https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/setting_sbp_cb.c#L365
+ Rust land `register_cb` digs out the context object and modifies the shared list of callbacks (again `setting_sbp_cb_register` should probably take the lock, not Rust)
+ We do the same thing, but for "ready by index done" message
   - It's not clear to be why we do this, in the main body of the function when we wait for the "index write response" message we don't match on the "done" message, it looks like it should just be treated as a match failure and we'd retry, see https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/settings.c#L171 and https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/request_state.c#L111 (not sure what this does actually, it might work because we unregister the "index read done" callback before it fires, because it looks like it tromps through all the outstanding requests and completes them: https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/setting_sbp_cb.c#L306)
+ We start the "send a message and wait for a response" process: https://github.com/swift-nav/libsettings/blob/master/src/settings.c#L601 
   - We build a request state object and append to a list in our shared state, the `request_state_append` correctly locks beforehand: https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/request_state.c#L131
   - We call `send_multi_thd` since we have a valid event object pointer
   - This goes into Rust land and calls `libsettings_send_from` which builds up an SBP message, and in the console eventually invokes: https://github.com/swift-nav/console_pp/blob/e7a97766a0cbaf62f7c874a599cc54e6a7fecfa2/console_backend/src/types.rs#L143
   - We move into wait, and pass in the event that Rust gave us
+ So eventually, the callback broker will invoke us from another thread (the receive thread in the console) -- we take the lock in Rust to find the callback then we drop the lock -- this this eventually calls https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/setting_sbp_cb.c#L248
  - This locks when looking up the request state from the context via: https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/request_state.c#L178
  - This triggers wake-up of the event object and the read thread here: https://github.com/swift-nav/libsettings/blob/9f89f1ad92864fe5a2d0c290b7d91f5a23c13e40/src/setting_sbp_cb.c#L291

Reply:
So I think the locking when accessing the shared state in Rust is probably correct for the callback broker.  I think we should add some locking with `LIBSETTINGS_LOCK` to `setting_sbp_cb_register` and `setting_sbp_cb_unregister`

Reply:
Suggested changes here: https://github.com/swift-nav/libsettings/pull/359